source/runtime.cpp

```diff
@@ -991,15 +1003,36 @@ void reshade::runtime::load_current_preset()
 	preset.get({}, "Techniques", technique_list);
 	std::vector<std::string> sorted_technique_list;
 	preset.get({}, "TechniqueSorting", sorted_technique_list);
-	std::vector<std::pair<std::string, std::string>> preset_preprocessor_definitions;
-	preset.get({}, "PreprocessorDefinitions", preset_preprocessor_definitions);
+
+	std::vector<definition> preprocessor_definitions = _preprocessor_definitions;
+
+	preprocessor_definitions.erase(std::remove_if(preprocessor_definitions.begin(), preprocessor_definitions.end(),
+		[](const definition &definition) { return definition.scope == definition::scope_preset || definition.scope == definition.scope_effect; }), preprocessor_definitions.end());
+
+	if (std::vector<std::pair<std::string, std::string >> preset_preprocessor_definitions;
+		preset.get({}, "PreprocessorDefinitions", preset_preprocessor_definitions))
+		for (std::pair<std::string, std::string> &value : preset_preprocessor_definitions)
+			preprocessor_definitions.push_back(definition{ definition::scope_preset, "", std::move(value.first), std::move(value.second) });
+
+	if (std::vector<std::string> effect_names; preset.get(effect_names))
+		for (std::string &effect_name : effect_names)
+			if (effect_name.rfind(".fx") != std::string::npos)
+				if (std::vector<std::pair<std::string, std::string >> effect_preprocessor_definitions;
+					preset.get(effect_name, "PreprocessorDefinitions", effect_preprocessor_definitions))
+					for (std::pair<std::string, std::string> &value : effect_preprocessor_definitions)
+						preprocessor_definitions.push_back(definition{ definition::scope_effect, effect_name, std::move(value.first), std::move(value.second) });
+
+	std::stable_sort(preprocessor_definitions.begin(), preprocessor_definitions.end(), [](const definition &a, const definition &b) { return a.first < b.first; });
+	std::stable_sort(preprocessor_definitions.begin(), preprocessor_definitions.end(), [](const definition &a, const definition &b) { return a.name < b.name; });
+	std::stable_sort(preprocessor_definitions.rbegin(), preprocessor_definitions.rend(), [](const definition &a, const definition &b) { return a.scope < b.scope; });
 
 	// Recompile effects if preprocessor definitions have changed or running in performance mode (in which case all preset values are compile-time constants)
 	if (_reload_remaining_effects != 0) // ... unless this is the 'load_current_preset' call in 'update_effects'
 	{
-		if (_performance_mode || preset_preprocessor_definitions != _preset_preprocessor_definitions)
+		if (_performance_mode || !std::equal(preprocessor_definitions.cbegin(), preprocessor_definitions.cend(), _preprocessor_definitions.cbegin(), _preprocessor_definitions.cend(),
+			[](const definition &a, const definition &b) { return a.scope == b.scope && a.name == b.name && a.first == b.first && a.second == b.second; }))
```

Comment:
It would be better to simply use the `operator!=` of the vector as before. Might have to implement a comparion operator for the `definition` struct, but since it's so simply I'd actually expect it to work out of the box.